pkg/tasker/tasker.go

```diff
@@ -0,0 +1,169 @@
+		CurrentChallenge: newVariables.Challenge,
+		Difficulty:       newVariables.Difficutly,
+		CurrentRequestId: newVariables.RequestIds,
+		TotalTips:        newVariables.Tip,
+	}, err
+}
+
+func (mt *Tasker) getNewChallengeChannel() (chan *tellor.ITellorNewChallenge, event.Subscription, error) {
+	sink := make(chan *tellor.ITellorNewChallenge)
+	var tellorFilterer *tellor.ITellorFilterer
+	tellorFilterer, err := tellor.NewITellorFilterer(mt.contractInstance.Address, mt.client)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "getting ITellorFilterer instance")
+	}
+	sub, err := tellorFilterer.WatchNewChallenge(&bind.WatchOpts{}, sink, nil)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "getting NewChallenge channel")
+	}
+	return sink, sub, nil
+}
+
+func (mt *Tasker) subscribeToNewChallenge() error {
+	var sink chan *tellor.ITellorNewChallenge
+	var err error
+	var sub event.Subscription
+	var failedSubscriptionCount int
+
+	// Initial subscription.
+	sink, sub, err = mt.getNewChallengeChannel()
+	if err != nil {
+		level.Error(mt.logger).Log("msg", "initial subscribing to NewChallenge events failed")
+		mt.resubscribe <- true
+	}
+	for {
+		select {
+		case <-mt.done:
+			if sub != nil {
+				sub.Unsubscribe()
+			}
+			level.Info(mt.logger).Log("msg", "unsubscribed to NewChallenge events")
+			return nil
+		case err := <-sub.Err():
+			if err != nil {
+				level.Error(mt.logger).Log(
+					"msg",
+					"new challenge subscription error",
+					"err", err)
+			}
+			mt.resubscribe <- true
+		case <-mt.resubscribe:
+			if failedSubscriptionCount == 10 {
+				return errors.New("failed to subscribe to NewChallenge events after 10 retries")
+			}
+			sink, sub, err = mt.getNewChallengeChannel()
+			if err != nil {
+				failedSubscriptionCount++
+				level.Error(mt.logger).Log("msg", "subscribing to NewChallenge events failed", "retry", failedSubscriptionCount)
+				mt.resubscribe <- true
+				continue
+			}
+			failedSubscriptionCount = 0
+			level.Info(mt.logger).Log("msg", "subscribed to NewChallenge events")
+		case vLog := <-sink:
+			mt.sendWork(vLog)
+		}
+	}
+}
+
+func (mt *Tasker) sendWork(challenge *tellor.ITellorNewChallenge) {
+	if challenge.CurrentRequestId[0].Int64() > int64(100) || challenge.CurrentRequestId[0].Int64() == 0 {
+		level.Warn(mt.logger).Log("msg", "new current variables request ID not correct - contract about to be upgraded")
+		return
+	}
+	newChallenge := &mining.MiningChallenge{
+		Challenge:  challenge.CurrentChallenge[:],
+		Difficulty: challenge.Difficulty,
+		RequestIDs: challenge.CurrentRequestId,
+	}
+
+	level.Debug(mt.logger).Log("msg", "new challenge for mining",
+		"hex", fmt.Sprintf("%x", newChallenge.Challenge),
+		"difficulty", newChallenge.Difficulty,
+		"requestIDs", fmt.Sprintf("%+v", newChallenge.RequestIDs),
+	)
+	mt.workSink <- &mining.Work{Challenge: newChallenge, PublicAddr: mt.accounts[0].Address.String(), Start: uint64(rand.Int63()), N: math.MaxInt64}
```

Comment:
I just realized that is it totally wrong.

The solution is specific to every public address to the tasker needs to send new work for each registered public address.

Reply:
And now this is quite tricky as the submitter needs to submit only the solution that corresponds to his public address.

Reply:
The solution is that each subscriber is indexed in a map like
```
submitters       [string]chan *Result
```

and then when you send the solution you send it to
```
ok,submitter := submitters[solution.Work.PublicAddr]

submitter <- solution
```